File: trunk/contrack/src/com/sinkluge/utilities/Verify.java
package com.sinkluge.utilities;

// Classes we will need
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Verify {

	public String message;

	public Verify () {
		message = "";
	}

	public static boolean blank(String s) {
		if (s ==null || s.equals("")) return true;
		else if (s.matches("^\\s+$")) return true;
		else return false;
	}

	public static boolean zip(String s) {
		 return s.matches("^\\d{5}([-]\\d{4})?$");
	}
	
	public static boolean phone(String s) {
		if (s != null && (
			s.matches("^(1-)?\\d{3}[-]\\d{3}[-]\\d{4}$") ||
			s.matches("^(1\\s)?\\(\\d{3}\\)\\s\\d{3}[-]\\d{4}$"))) return true;
		else return false;
	}

	public static boolean email(String s) {
		if (s != null && s.matches("^[\\w-]+(\\.([\\w-]+))*@[\\w-]+(\\.([\\w-]+))*\\.[a-z]+$")) return true;
		else return false;
	}

	public static boolean date(String s) {
		if (s == null) return true;
		boolean valid = true;
		if (s.matches("^\\d{2}/\\d{2}/\\d{4}$")) {
			try {
				int year = Integer.parseInt(s.substring(6));
				int month = Integer.parseInt(s.substring(0,2));
				int day = Integer.parseInt(s.substring(3,5));
				if (day > 31 || month < 1 || month > 12) valid = false;
				else if ((month == 4 || month == 6 || month == 9 || month == 11) && day > 30) valid = false;
				else if (month == 2) {
					if (year % 4 == 0 && day > 29) valid = false;
					else if (day > 28) valid = false;
				}
			} catch (Exception e) {
				valid = false;
			}
		} else valid = false;
		return valid;
	}

	public static boolean integer(String s) {
		try {
			if (blank(s)) s = "0";
			Integer.parseInt(s);
			return true;
		} catch (NumberFormatException e) {
			return false;
		}
	}

	public static boolean cur(String s) {
		try {
			if (blank(s)) s = "0";
			Double.parseDouble(s);
			return true;
		} catch (NumberFormatException e) {
			return false;
		}
	}

	public void msg(String append) {
		if (!blank(message)) message += ", ";
		message += append;
	}

	public boolean password (String s, String s2) {
		if (!blank(s)) {
			if (s2 != null && !s.equals(s2)){
				msg("Passwords do not match");
				return false;
			} else if (s.length() < 6) {
				msg("Passwords must contain 6 characters");
				return false;
			} else {
				Matcher a = Pattern.compile("[a-zA-Z]").matcher(s);
				Matcher ns = Pattern.compile("[\\d\\W]").matcher(s);
				if (!a.find() || !ns.find()) {
					msg("Passwords must contain at least one letter and one number or symbol");
					return false;
				}
			}
		} else {
			msg("Password cannot be empty");
			return false;
		}
		return true;
	}
}

File: trunk/external/WebContent/jsp/scripts/verify.js
function req(obj) {
	return (obj.value == null || obj.value == "" || isBlank(obj));
}
function isBlank (obj) {
	if(obj.value.match(/^\s+$/)) return true;

	else return false;
}	
function vDate(obj) {
	if (!obj.value.match(/^\d{2}\/\d{2}\/\d{4}$/)) return false;

	var year = obj.value.substring(6);
	var month = obj.value.substring(0,2);
	var day = obj.value.substring(3,5);
	var year = year-0;
	var month = month-0;
	var day = day-0;
	if (day > 31 || month < 1 || month > 12) return false;
	else if ((month == 4 || month == 6 || month == 9 || month ==11) && day > 30) return false;
	else if (month == 2) {
		if (year % 4 == 0 && day > 29) return false;
		else if (day > 28) return false;
	}
	return true;
}
function vInt(s) {
	deComma(s);
	if(s.value.match(/^([-]?)(\d*)$/)) return true;
	else return false;
}
function vFloat (s){
	deComma(s);
	if(s.value.match(/^([-]?)(\d*)(\.?)(\d*)$/)) return true;
	else return false;
}
function deComma (s) {
	for (var i = 0; i < s.value.length; i++) {
		var c = s.value.charAt(i);
		if (c == ",") s.value = s.value.substring(0,i) + s.value.substring(i+1);
	}
}

// Try to get rid of this function
function isDigit(c) {
	return ((c >= "0") &&  (c <= "9"));
}
function vEmail(s) {
	if (s.value.match(/^([a-zA-Z0-9])+([.a-zA-Z0-9_-])*@([a-zA-Z0-9_-])+(.[a-zA-Z0-9_-]+)+/)) return true;
	else return false;	
}
function vPhone(s) {
	if (s.value.match(/^\(\d{3}\)\s\d{3}[-]\d{4}$/)) return true;
	else {
		if (s.value.match(/^\d{3}[-]\d{3}[-]\d{4}$/)) {
			s.value = "(" + s.value.substring(0,3) + ") " + s.value.substring(4,7) + "-" + s.value.substring(8);
			return true;
		}
		else return false;
	}
}
function vZip(s) {
	if (s.value.match(/^\d{5}([-]\d{4})?$/)) return true;
	else return false;
}
function vPassword(s) {
	if (s.currentPassword != null && req(s.currentPassword)) {
		errors += "- Current password is required";
		s.currentPassword.style.backgroundColor = "#FFFFCC";
	} else if (s.value != s.passwordMatch.value) {
		errors += "- Passwords do not match.\n"
		s.style.backgroundColor = "#FFFFCC";
	} else if (!s.value.match(/[\d\W]/) || !s.value.match(/[a-zA-Z]/)) {
		errors += "- Password must contain at least one alphabetic character and one number or symbol\n";
		s.style.backgroundColor = "#FFFFCC";
	} else if (s.value.length < 6) {
		errors += "- Passwords must be at least six characters long.\n";
		s.style.backgroundColor = "#FFFFCC";
	}
}
var errors = "";
function checkForm(obj) {
	var msg;
	var empty_fields = "";
	errors = "";
	for (var i = 0; i < obj.length; i++) {
		var e = obj.elements[i];
		if (!e.disabled) {
			if (e.type != "button" && e.type != "submit") e.style.backgroundColor = "white";
			if(e.required && req(e)) {
				empty_fields += "\n\t" + e.eName;
				e.style.backgroundColor = "#FFFFCC";	
			} else if(e.isInt && !vInt(e) && !req(e)) {
				errors += "- " + e.eName + " contains an invalid integer number.\n";
				e.style.backgroundColor = "#FFFFCC";
			} else if(e.isFloat && !vFloat(e) && !req(e)) {
				errors += "- " + e.eName + " contains an invalid decimal number.\n"; 
				e.style.backgroundColor = "#FFFFCC";
			} else if(e.isDate && !vDate(e) && !req(e)) {
				errors += "- " + e.eName + " contains an invalid date.\n";
				e.style.backgroundColor = "#FFFFCC";
			} else if(e.isEmail && !vEmail(e) && !req(e)) {
				errors += "- " + e.eName + " contains an invalid email address.\n";
				e.style.backgroundColor = "#FFFFCC";
			} else if(e.isPhone && !vPhone(e) && !req(e)) {
				errors += "- " + e.eName + " contains an invalid phone number.\n\tExample: ###-###-####.\n";
				e.style.backgroundColor = "#FFFFCC";
			} else if(e.isZip && !vZip(e) && !req(e)) {
				errors += "- " + e.eName + " contains an invalid zip code.\n\tExample: ##### or #####-####.\n";
				e.style.backgroundColor = "#FFFFCC";
			} else if(e.isPassword) vPassword(e);
		} else e.style.backgroundColor = "#c0c0c0";
	}

	//Are there errors?
	if (!empty_fields && !errors) return true;
	
	//Display errors
	msg  = "_________________________________________________\n\n";
	msg += "The form was not submitted because of the following error(s).\n";
	msg += "Please correct these error(s) and re-submit.\n";
	msg += "_________________________________________________\n\n";
	
	if (empty_fields) {
		msg += "- The following required field(s) are empty:" + empty_fields + "\n";
		if (errors) msg += "\n";
	}
	
	msg += errors;
	alert(msg);
	return false;
}